### PR TITLE
DOC L1-regularization parameter tutorial

### PR DESCRIPTION
**Pull request by wassimmazouz**
This PR aims to add documentation about the mathematical context behind the choice of L1-regularization in `skglm` as it is one of the frequently asked questions.

**Tasks :** 
- [x] Created `.rst` tutorial and added its tag to the `tutorials.rst` file
- [x] Introduced the problem and aim of the tutorial
- [x] Finished the theoretical background
- [x] Illustrated with an example
- [x] Added references

### CODE REVIEW

**Comment by Badr-MOUFAD** [comment]
Thanks for the PR @wassimmazouz

I will follow along with you so ping me when you are done or need a hand

File: doc/tutorials/fermat_rule_reg.rst [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+=============
+
+Consider the optimization problem:
+
+.. math::
+    \min_x f(x) + \lambda \|x\|_1
+
+where:
+
+- :math:`f: \mathbb{R}^d \to \mathbb{R}` is a differentiable function,
```
Comment (mathurinm) [inline doc/tutorials/fermat_rule_reg.rst:24]: Add convex

File: doc/tutorials/fermat_rule_reg.rst [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+According to Fermat's rule, the minimum of the function occurs where the subdifferential of the objective function includes zero. For our problem, the objective function is:
+
+.. math::
+    g(x) = f(x) + \lambda \|x\|_1
+
+The subdifferential of :math:`\|x\|_1` at 0 is the L-infinity ball:
+
+.. math::
+    \partial \|x\|_1 |_{x=0} = \{ u \in \mathbb{R}^n : \|u\|_{\infty} \leq 1 \}
```
Comment (mathurinm) [inline doc/tutorials/fermat_rule_reg.rst:41]: We use Rd rather than Rn, as above

File: doc/tutorials/fermat_rule_reg.rst [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+Thus, for :math:`0 \in \partial g(x)` at :math:`x=0`:
+
+.. math::
+    0 \in \nabla f(0) + \lambda \partial \|x\|_1 |_{x=0}
+
+which implies, given that the dual of L1-norm is L-infinity:
```
Comment (mathurinm) [inline doc/tutorials/fermat_rule_reg.rst:48]: Precise what you lean here because dual has many meanings

File: doc/tutorials/fermat_rule_reg.rst [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+.. math::
+    \|\nabla f(0)\|_{\infty} \leq \lambda
+
+If :math:`\lambda > \|\nabla f(0)\|_{\infty}`, then the only solution is :math:`x=0`.
+
+Example
+=======
+
+Consider the loss function for Ordinary Least Squares :math:`f(x) = \frac{1}{2} \|Ax - b\|_2^2`. We have:
```
Comment (mathurinm) [inline doc/tutorials/fermat_rule_reg.rst:58]: Good idea. Minor adaptation: we use a normalized datfit, divided by n samples

File: doc/tutorials/tutorials.rst [hunk 5]
```diff
@@ -33,3 +33,8 @@ Get details about Cox datafit equations.
 -----------------------------------------------------------------
 
 Mathematical details about the group Lasso, in particular with nonnegativity constraints.
+
+:ref:`Mathematics Behind L1 Regularization and Fermat's Rule <fermat_rule_reg>`
```
Comment (mathurinm) [inline doc/tutorials/tutorials.rst:37]: I'd rename this critical régularisation strength above which solution is 0 (less maths, more to the point)

File: doc/tutorials/fermat_rule_reg.rst [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+======================================================
+Mathematics Behind L1 Regularization and Fermat's Rule
+======================================================
+
+This tutorial presents the mathematics behind solving the optimization problem
+:math:`\min f(x) + \lambda \|x\|_1` and demonstrates why the solution is zero when
+:math:`\lambda` is greater than the infinity norm of the gradient of :math:`f` at zero, therefore justifying the choice in skglm of
+
+.. code-block::
```
Comment (mathurinm) [inline doc/tutorials/fermat_rule_reg.rst:11]: Add a sentence saying that parametrizing the régularisation as a fraction of this, eg 0.01 alpha max, is more convenienti

**Comment by mathurinm** [comment]
@wassimmazouz almost done, can you look into this rendering bug for the align environment that I was not able to fix ?



image

**Comment by wassimmazouz** [comment]
@mathurinm Is it a local issue? It renders correctly for me
image

**Comment by mathurinm** [comment]
We have an action that builds the doc on the PR and publishes it (`build doc artifacts` on circle CI), see the buggy rendering here too: https://output.circle-artifacts.com/output/job/c0047b51-706d-4b6c-ba96-4160db844411/artifacts/0/dev/tutorials/alpha_max.html#theoretical-background (from 1 commit ago)

**Comment by wassimmazouz** [comment]
@mathurinm it might have to do with the `sphinx` or `mathjax` versions. Some forums report similar errors in older versions of these packages. I tried to change the version and test locally but no progress for now. 
I added a `eqnarray` to try if it works without changing anything else.

**Comment by Badr-MOUFAD** [comment]
The problem seems to appear only for multiline equations.
Fixed it by adding ``:nowrap:`` directive and writing plain tex.

Perhaps, sphinx/mathjax has troubles parsing the equation 🤷‍♂️

**Comment by mathurinm** [comment]
@wassimmazouz can you report this back to sphinx with a minimal example ?

**Review by mathurinm — APPROVED** [review]
Looking great, thanks @wassimmazouz @Badr-MOUFAD

**Pull request merged**